File: src/main/java/ting/leetcode/java/arrays101/SquaresOfASortedArray.java
package ting.leetcode.java.arrays101;

public class SquaresOfASortedArray {
    public int[] sortedSquares(int[] nums) {
        final int[] results = new int[nums.length];
        int headIndex = 0;
        int tailIndex = nums.length - 1;
        int cursor = nums.length - 1;
        while (tailIndex >= headIndex) {
            int tailElement = Math.abs(nums[tailIndex]);
            int headElement = Math.abs(nums[headIndex]);
            if (headElement > tailElement) {
                results[cursor] = (int) Math.pow(headElement, 2) ;
                headIndex++;
            } else {
                results[cursor] = (int) Math.pow(tailElement, 2);
                tailIndex--;
            }
            cursor--;
        }
        return results;
    }

    public int[] sortedSquares2(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        int[] ans = new int[nums.length];
        int index = nums.length - 1;


        while (left <= right) {
            int rightPow = (int) Math.pow(nums[right], 2);
            int leftPow = (int) Math.pow(nums[left], 2);

            if (rightPow >= leftPow) {
                ans[index] = rightPow;
                right--;
            } else {
                ans[index] = leftPow;
                left++;
            }
            index--;
        }

        return ans;
    }
}

File: src/main/kotlin/ting/leetcode/arrays101/SquaresOfASortedArray.kt
package ting.leetcode.arrays101

class SquaresOfASortedArray {
    fun sortedSquares(nums: IntArray): IntArray {
        var left = 0
        var right = nums.lastIndex
        val ans = IntArray(nums.size)

        for (i in ans.lastIndex downTo 0) {
            val leftPow = nums[left] * nums[left]
            val rightPow = nums[right] * nums[right]
            if (rightPow >= leftPow)
                ans[i] = rightPow.also { right-- }
            else
                ans[i] = leftPow.also { left++ }
        }
        return ans
    }
}
